src/main/java/net/minecraftforge/client/model/ModelFluid.java

```diff
@@ -170,158 +170,232 @@ public BakedFluid(Optional<TRSRTransformation> transformation, ImmutableMap<Tran
+                {
+                    key <<= 10;
+                    key |= cornerRound[i];
+                }
+                key <<= 1;
+                key |= 1;
+
+                return modelCache.getUnchecked(key).getQuads(state, side, rand);
+            }
+
+            return super.getQuads(state, side, rand);
+        }
+    }
+
+    private static class BakedFluid implements IBakedModel
+    {
+        private static final int x[] = { 0, 0, 1, 1 };
+        private static final int z[] = { 0, 1, 1, 0 };
+        private static final float eps = 1e-3f;
+
+        protected final Optional<TRSRTransformation> transformation;
+        protected final ImmutableMap<TransformType, TRSRTransformation> transforms;
+        protected final VertexFormat format;
+        protected final int color;
+        protected final TextureAtlasSprite still, flowing;
+        protected final Optional<TextureAtlasSprite> overlay;
+        protected final boolean gas;
+        protected final ImmutableMap<EnumFacing, ImmutableList<BakedQuad>> faceQuads;
+
+        public BakedFluid(Optional<TRSRTransformation> transformation, ImmutableMap<TransformType, TRSRTransformation> transforms, VertexFormat format, int color, TextureAtlasSprite still, TextureAtlasSprite flowing, Optional<TextureAtlasSprite> overlay, boolean gas, boolean statePresent, int[] cornerRound, int flowRound, boolean[] sideOverlays)
         {
             this.transformation = transformation;
             this.transforms = transforms;
             this.format = format;
             this.color = color;
             this.still = still;
             this.flowing = flowing;
+            this.overlay = overlay;
             this.gas = gas;
+            this.faceQuads = buildQuads(statePresent, cornerRound, flowRound, sideOverlays);
+        }
 
-            faceQuads = Maps.newEnumMap(EnumFacing.class);
-            for(EnumFacing side : EnumFacing.values())
+        private ImmutableMap<EnumFacing, ImmutableList<BakedQuad>> buildQuads(boolean statePresent, int[] cornerRound, int flowRound, boolean[] sideOverlays)
+        {
+            EnumMap<EnumFacing, ImmutableList<BakedQuad>> faceQuads = new EnumMap<>(EnumFacing.class);
+            for (EnumFacing side : EnumFacing.values())
             {
                 faceQuads.put(side, ImmutableList.of());
             }
 
-            if(statePresent)
+            if (statePresent)
             {
+                // y levels
                 float[] y = new float[4];
-                for(int i = 0; i < 4; i++)
+                boolean fullVolume = true;
+                for (int i = 0; i < 4; i++)
                 {
-                    if(gas)
-                    {
-                        y[i] = 1 - cornerRound[i] / 768f;
-                    }
-                    else
-                    {
-                        y[i] = cornerRound[i] / 768f;
-                    }
+                    float value = cornerRound[i] / 864f;
```

Comment:
More magic numbers, I understand fry used a bunch before hand but still. Probably better to document them for the next person along.